File: recursion/recursion01.java
import java.util.Scanner;
import java.util.ArrayList;

public class recursion01 {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        solve();
    }
    
    static void solve() {
        // basics();
        questionSet();
    }

    static void basics() {
        int a = scn.nextInt();
        int b = scn.nextInt();
        System.out.println(fact(a));
        System.out.println(powerBtr(a, b));

    }

    public static int fact(int n) {
        if(n <= 1) return 1;
        return fact(n - 1) * n;
    }

    public static int powerBtr(int a, int b) {
        if(b == 0)  return 1;
        int halfPwr = powerBtr(a, b / 2);
        halfPwr *= halfPwr;
        return halfPwr * (b % 2 == 0 ? 1 : a);
    }

    public static int firstIndex(int[] arr, int idx, int data) {
        if(arr.length == idx)   return -1;
        if(arr[idx] == data) {
            return idx;
        } 
        return firstIndex(arr, idx + 1, data);
    }

    public static int lastIndex(int[] arr, int idx, int data) {
        if(arr.length == idx)   return -1;
        int lisa = lastIndex(arr, idx + 1, data);
        if(lisa != -1) {
            return idx;
        } else {
            if(arr[idx] == data){
                return idx;
            }   else {
                return -1;
            }
        }
    }

    public static int[] allIndexes(int[] arr, int idx, int data, int count) {
        if(arr.length == idx) {
            int[] ans = new int[count];
            return ans;
        }

        if(arr[idx] == data)    count++;
        int[] ans = allIndexes(arr, idx + 1, data, count);
        if(arr[idx] == data) {
            ans[count - 1]  = idx;
        }
        return ans;
    }


    // questionSet()========================
    
    public static void questionSet() {
        // System.out.println(permutation("abcd"));
        // System.out.println(substr("abc", ""));
        // System.out.println(permutation("aba", ""));
        System.out.println(permutationUnique("aba", ""));
    }

    public static ArrayList<String> subsequence(String str) {
        if(str.length() == 0)   {
            ArrayList<String> base = new ArrayList<>();
            base.add("");
            return base;
        }
        char ch = str.charAt(0);
        String nstr = str.substring(1);

        ArrayList<String> smallAns = subsequence(nstr);
        ArrayList<String> myAns = new ArrayList<>();

        for(String s : smallAns) {
            myAns.add(s);
            myAns.add(ch + s);
        }

        return myAns;
    } 

    public static ArrayList<String> permutation(String str) {
        if(str.length() == 1){
            ArrayList<String> bres = new ArrayList<>();
            bres.add(str);
            return bres;
        }

        char ch = str.charAt(0);
        String nstr = str.substring(1);

        ArrayList<String> rres = permutation(nstr);
        ArrayList<String> myAns = new ArrayList<>();

        for(String s : rres) {
            for(int i = 0; i <= s.length(); i++){
                myAns.add(s.substring(0, i) + ch + s.substring(i));
            }
        }

        return myAns;
    }

    public static int substr(String str, String ans) {
        if(str.length() == 0) {
            System.out.println(ans);
            return 1;
        }

        char ch = str.charAt(0);
        String nstr = str.substring(1);
        int count = 0;

        count += substr(nstr, ans + ch);
        count += substr(nstr, ans);

        return count;
    }

    public static int permutation(String str, String ans) {
        if(str.equals("")){     
            System.out.println(ans);
            return 1;
        }
        int count = 0;
        for(int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            String nstr = str.substring(0, i) + str.substring(i + 1);
            count += permutation(nstr, ch + ans);
        }
        return count;
    }

    public static int permutationUnique(String str,String ans){
        if(str.length()==0){
            System.out.println(ans);
            return 1;
        }

        int count=0;
        boolean[] vis=new boolean[26];
        for(int i=0;i<str.length();i++){
            char ch=str.charAt(i);
            if(!vis[ch-'a']){
                vis[ch-'a']=true;
                String nstr=str.substring(0,i) + str.substring(i+1);
                count+=permutationUnique(nstr,ch + ans);
            }
        }
        return count;
    }
}

File: recursion/recursion01.cpp
#include <iostream>
#include <vector>
using namespace std;

void printInc(int a, int b) {
    if(a == b + 1)      
        return;
    cout << a << ", ";
    printInc(a + 1, b);
}

void printDec(int a, int b) {
    if(a == b + 1)      
        return;
    printDec(a + 1, b);
    cout << a << ", ";
}

void printIncDec(int a, int b) {
    if(a == b + 1)      
        return;
    if(a % 2 == 0)
    cout << a << " hi" << endl;
    printIncDec(a + 1, b);
    if(a % 2 != 0)
    cout << a << " bye" << endl;
}



void recursionWithArray() {
    vector<int> arr{10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
    cout << maximum(arr, 0);
}

int maximum(vector<int>& arr, int idx) {
    if(idx == arr.size() - 1)   return arr[idx];
    int max1 = arr[idx];
    return max(max1, maximum(arr, idx + 1));
}

// questionSet()========================


vector<string> words = {
    ":;/",
    "abc",
    "def",
    "ghi",
    "jkl",
    "mno",
    "pqrs",
    "tuv",
    "wx",
    "yz",
    "&*%",
    "#@$",
};

vector<string> keypad(string str) {
    if(str.length == 0) {
        vector<string> base;
        base.push_back("");
        return base;
    }

    char ch = str[0];
    string nstr = str.substr(1);
    int num = ch - '0';
    string word = words[num];
    vector<string>  smallAns = keypad(nstr);
    vector<string> myAns;

    for(string s : smallAns) {
        for(int i = 0; i < word.length; i++){
            myAns.push_back(word[i] + s);
        }
    }
    return myAns;
}



void questionSet() {
    vector<string> ans = keypad("489");
    for(string s : ans){
        cout << s << " ";
    }
}

void basics() {
    int a, b;
    cin >> a >> b;
    // printInc(a, b);
    // printDec(a, b);
    printIncDec(a, b);
}


void solve() {
    // basics();
    // recursionWithArray();
    questionSet();
}

int main() {
    solve();
    return 0;
}